[WIP] #46970 Milestone 2

oisincoveney opened the pull request.

WiNloSt committed:
Add a new SDK setting page and it's dependencies (#47189)

* Add a new SDK setting page and it's dependencies

* Migrate to use `component` for admin settings

* [WIP] add an SDK setting using component API

* Simplify admin code

* Refactor the use of `getSettings`

* Make updating admin setting easier

* Add a new `embedding-app-origins-sdk` setting

* Fix embedding card and copy

* Fix static and interactive embedding icon not using custom brand color

* Update embedding cards copy

* Fig Figma reference link

* Fix disabled text and placeholder color

* Disable SDK origins according to a condition

* Fix type error

* Fix unit tests 1/*

* Fix unit tests 2/2

* Fix unit tests 3/3

* Fix BE tests

* Revert the copy and order of embedding cards back

* Fix E2E tests

* Fix embedding SDK breadcremb

* Revert "Make updating admin setting easier"

This reverts commit 46506ede91442aa5dd2e6e0e0151e23e440ea958.

* Revert "Refactor the use of `getSettings`"

This reverts commit b644928a4181fce4b2db5c4127f470d738f64227.

* Revert "Simplify admin code"

This reverts commit 14b4ffa4a1d97054e8fffde7ba3181d33d962367.

* Revert all settings changes to not interfere with the new project

* Remove unused export

* Update color mapping

WiNloSt committed:
Merge branch 'master' into 46970-milestone-1

WiNloSt committed:
Add embedding SDK admin settings tests (#47519)

* Add static embed tests

* Consolidate embedding tests

* Consolidate embedding tests

* Consolidate interactive embed unit tests

* Add embedding SDK card tests

* Simplify tests

* Add embedding SDK tests

* Share the same repeated types

* Fix test name typo

WiNloSt committed:
Merge branch 'master' into 46970-milestone-1

WiNloSt committed:
Merge branch 'master' into 46970-milestone-1

WiNloSt committed:
Merge branch 'master' into 46970-milestone-1

WiNloSt committed:
Introduce `enable-embedding-sdk` toggle (#47686)

* pass `embedding-sdk` flag in `token-features`

* Add the embedding SDK toggle to the SDK card

* Add embedding SDK toggle to its settings page

* Enable the CORS origin input based on the SDK enable setting

* Fix BE tests

* Correct token feature `embedding_sdk` name

* fix unit tests

* Add embedding SDk toggle tests

* Add embedding SDK settings page toggle tests

* Hook up the new settingss, feature flag together

- feature flag `embedding-sdk`
- setting `enable-embedding-sdk`
- setting `embedding-app-origins-sdk`

WiNloSt committed:
Merge branch 'master' into 46970-milestone-1

WiNloSt committed:
Add embedding SDK CORS tests (#47761)

* Add E2E CORS tests

* Test CORS with `embedding-app-origins-sdk`

* Make custom URL consistent

* Simplify the code, not introducing more variants to package.json

WiNloSt committed:
Merge branch 'master' into 46970-milestone-1

WiNloSt committed:
Merge branch 'master' into 46970-milestone-1

WiNloSt committed:
Add 2 more embedding toggle settings (#47770)

* Add 2 more embedding toggle settings

- static embedding
- interactive embedding

* Add 2 more embedding toggles to embedding settings page

* Makes spacing easier to understand

* Add static embed settings page

* Add interactive embedding settings page

* Make unit tests consistent

* Add missing static embedding assertions

* Fix import dependency causing EE plugins to load after it should

* Fix unit tests

* Fix embedding-smoketests.cy.*

* Update E2E snapshots following the previous logic

* Fix more E2E tests

* Revert BE change becuase it fails a tons of BE tests.

* Make git history cleaner

* Clarify todo comment

WiNloSt committed:
Respect `embedding` feature flag for interactive toggle (#47883)

* Respect `embedding` feature flag for interactive toggle

* Fix settings inside interactive embedding not working

Commit:
Merge branch 'master' into 46970-milestone-2

Commit:
Merge branch 'master' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

WiNloSt committed:
Update embedding admin settings copy (#47891)

Co-authored-by: Oisin Coveney <oisin@metabase.com>

Commit:
Merge branch 'master' into 46970-milestone-2

Commit:
Merge branch 'master' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
set encryption:no for embedding-app-origin as a temp? fix

escherize committed:
Feature flag and settings for SDK part one (#47933)

Co-authored-by: Oisin Coveney <oisin@metabase.com>

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

npretto committed:
make tests pass on 46970-milestone-2 (#48293)

* make embedding-smowktests pass by setting enable-embedding-static to true in the snapshot, and by moving to that setting when checking if static is enabled

* revert validation check for embedding

* TO BE REVERTED: set enable-embedding when enable-embedding-static is set until we have the BE migrated

* update `public-sharing-embed-button-behavior.cy.spec.js` to use `enable-embedding-static` instead of `enable-embedding`

* update sharing menu test setup to use new setting name

---------

Co-authored-by: Oisin Coveney <oisin@metabase.com>

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Add snowflow events for embedding usage (#48305)

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

WiNloSt committed:
Handle settings set via environment variables (#47929)

Co-authored-by: Oisin Coveney <oisin@metabase.com>

Somtom commented on snowplow/iglu-client-embedded/schemas/com.metabase/embed_share/jsonschema/1-0-1:
**question:** Have we considered using the new [simple events](https://www.notion.so/metabase/Snowplow-integration-5da1f874beda4153b4fccfa6c1e77caa?pvs=4#59fb20748e3248ab832d62b9dc981a15) to track those interactions?

The approach you take here definitely would work. However, the downside is that we need this schema change (which also needs to be populated to SnowcatCloud later in order to work + we need Joao from SnowcatCloud to help us). Also, every time there is a new event, we need to update the schema version again (just because of the enum).

Thus, I was wondering whether we could also use the simple events approach instead. This only really works, if the additional data like `number_embedded_questions`, etc. are not needed. The simple events [schema](https://github.com/metabase/metabase/blob/master/snowplow/iglu-client-embedded/schemas/com.metabase/simple_event/jsonschema/1-0-0#L51) have a `event_detail` field which can be used to send some additional information but it would not fit for all 3 contexts we send in this schema.

[Here](https://github.com/metabase/metabase/pull/47410/files#diff-43bf5c088b93a140e2a01fb0e70ca7abf9ddd9c7f24c4b55d4566166a194f593R20-R24) is an example how you would use the simple event.

escherize committed:
feat(sdk): adds granular embedding settings (#48154)

* Resolves many issue for milestone 0 of Feature flag and settings for SDK epic

- Add a new boolean setting `enable-embedding-sdk`
- Add a new string setting `embedding-app-origins-sdk`
- Add a new boolean setting `enable-embedding-static`
- Add a new boolean setting `enable-embedding-interactive`

- Depricate `enable-embedding`

- Also hook up localhost behaviors for embedding-app-origins-sdk

- Setup embedding toggle analytics for enable-embedding-{'sdk','static','interactive',''}

- hook up the embedding-app-origi{'n','ns-sdk'} in server.middleware/security
- bit of cleanup in server.middleware/security

* remove print statement

* indent + lint

* need to have :embedding and :embedding-sdk

* indent

* Don't allow setting embedding-app-origins-sdk without feature

- But you should be able to read from it

* `=` -> `= `

* fixup tests

* more test fixing

* fix the last test + add comment

* use with-temporary-setting-values

* review responses + add more docs

* Add test for deprecated setting warning + one for embedding

* Add encryption to embedding-app-origin

* Change encryption:never to encryption:no

* updates for new encryption values + fix merge

* cljfmt appeasement

* Fix metabase.server.middleware.security-test/test-access-control-headers?

* add comment + take a swing at e2e test

* first crack at part-two of milestone0

* check and sync deprecated X supported embedding settings

- with tests
- some merge fixes (remove duplicate defsettings that snuck in)

* hook in check-and-sync-settings

* fixup merge conflict naming

* indent

* indent

* migration should work on mysql+h2+pg

* finish tests for migrations

* linter fix

* fix embed/settings-test

* fix a ton of tests

- enable-embedding-* shouldn't need the embedding feature, just like
  enable-embedding doesnt

* fix 6 more tests

* JWT depends on enable-embedding-static.

* fix tests

- w/ enable-embedding -> enbable-embedding-static

* more embedding settings in test tweaks

* fixing more tests

* uncomment test +

- fix feature config for enable-embedding-interactive
- fix feature config for embedding-app-origins-interactive

* jwt => static, not interactive.

* tracking down and fixing littany of test failures

* fix embed.settings-test

* fix another

- `enable-embedding` -> `enable-embedding-static`

* fix about 30 tests

* add tests for migrating encrypted columns

* fix MORE tests

* fix api/dashboard_test + typo

* indenation

* fix typo in migrations

* pass in env

* be more explicit about settings in `test-access-control-headers?`

- improve test that takes multiple origins (moved the matching origin to
  the end)

* be more explicit about embedding settings in `sync-origins-test`

* finally fix embed.settings

* need to look for :mb-<setting> in env, not :<setting>

* fix access-control-headers tests.

* clarify what's required for `test-access-control-headers?`

add bangs to embedding settings checking and syncing

* remove flaky snowplow test event + linter

* dont setTokenFeatures in embedding smoketests.

* embedding-app-origins-sdk-always-allows-localhost-test

* sdk embedding requires the :embedding-sdk feature.

* fix test-access-control-headers? test

* Localhost is always permitted for sdk tinkering

+ better checks for embedding-app-origin-set

* actually, localhost is only permitted when `enable-embedding-sdk` is `true`.

* Only set embedding origins when we have features

* fix test-access-control-headers

* setting embedding-app-origins-sdk requires :embedding-sdk feature

* fix embedding-smoketests e2e

- enable-embedding -> enable-embedding-static

* don't migrate embedding-app-origin -> embedding-app-origins-sdk (as per product doc)

* don't migrate embedding-app-origin -> embedding-app-origins-sdk (re: product doc)

* update docstring in test helper

---------

Co-authored-by: Oisin Coveney <oisin@metabase.com>

Commit:
Merge branch 'sdk-launch-in-core-app' into 46970-milestone-2

Commit:
Revert `enable-embedding` workarounds in Milestone 2 (#48435)
